apps/prairielearn/src/sprocs/ip_to_mode.sql

```diff
@@ -22,63 +20,66 @@ BEGIN
         mode := 'Public';
     END IF;
 
-    -- Does the user have an active PT reservation?
-    SELECT r.session_id
-    INTO v_session_id
-    FROM
-        pt_reservations AS r
-        JOIN pt_enrollments AS e ON (e.id = r.enrollment_id)
-    WHERE
-        e.user_id = ip_to_mode.authn_user_id
-        AND (date BETWEEN r.access_start AND r.access_end);
-
-    IF FOUND THEN
-        -- The user has a checked-in pt_reservation. Check whether the
-        -- reservation is in a testing center location that requires network
-        -- filtering.
-
-        SELECT loc.id, loc.filter_networks
-        INTO v_location_id, v_filter_networks
+    -- Consider each PT reservation which is either active or corresponds to
+    -- a session that will start soon or started recently.
+    FOR reservation IN
+        SELECT
+            r.session_id,
+            (ip_to_mode.date BETWEEN r.access_start AND access_end) AS reservation_active,
+            l.id AS location_id,
+            l.filter_networks AS location_filter_networks
         FROM
-            pt_sessions AS s
-            JOIN pt_locations AS loc ON (loc.id = s.location_id)
+            pt_reservations AS r
+            JOIN pt_enrollments AS e ON (e.id = r.enrollment_id)
+            JOIN pt_sessions AS s ON (s.id = r.session_id)
+            LEFT JOIN pt_locations AS l ON (l.id = s.location_id)
         WHERE
-            s.id = v_session_id;
-
-        IF FOUND THEN
-            -- The reservation is in a testing center location.
-            IF v_filter_networks THEN
-                -- The reservation is in a testing center location that requires
-                -- network filtering.
-
-                PERFORM *
-                FROM pt_location_networks AS ln
-                WHERE
-                    ln.location_id = v_location_id
-                    AND ip <<= ln.network;
-
-                IF FOUND THEN
-                    -- The user is physically inside the testing center. Set
-                    -- mode to 'Exam'.
-                    mode := 'Exam';
-                ELSE
-                    -- Although we have a checked-in reservation, the user is
-                    -- physically outside the testing center. Set mode to
-                    -- 'Public'.
-                    mode := 'Public';
-                END IF;
-            ELSE
-                -- The reservation is in a testing center location that doesn't
-                -- require network filtering. Set mode to 'Exam'.
+            e.user_id = ip_to_mode.authn_user_id
+            AND (
+                (r.access_end IS NULL and s.date BETWEEN ip_to_mode.date - '1 hour'::interval and ip_to_mode.date + '1 hour'::interval)
```

Comment:
Nit: when comparing the current time (in this case, `ip_to_mode.date`) to an offset dates, I normally write it the other way around. In this particular case it doesn't really matter because we are doing +/- 1 hour, but if we have asymmetric intervals then I find a requirement like "from 1 hour before session start to 30 minutes after session start" to be clearer when written:

`ip_to_mode.date BETWEEN s.date - '1 hour'::interval AND s.date + '30 minutes'::interval`

rather than the equivalent way we'd write it as above:

`s.date BETWEEN ip_to_mode.date - '30 minutes'::interval AND ip_to_mode.date + '1 hour'::interval`

(Note the reversal of the signs on the 30min/1hour intervals).

Reply:
That makes a lot of sense, I probably would have noticed this if they weren't symmetrical, like you said!